### PR TITLE
SYNCT-252: Added sync2 hash code generation for openmrs's objects

### PR DESCRIPTION
**Pull request by wojtekbu**
* Added unit tests

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/sync2/api/utils/SyncHashcodeUtils.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+		removeVoided(result, clazz);
+		removeId(result, clazz);
+		return result;
+	}
+
+	private static void removeVoided(final SimpleObject result, Class<?> clazz) {
+		if (BaseOpenmrsData.class.isAssignableFrom(clazz)) {
+			result.remove("dateVoided");
+			result.remove("voided");
+			result.remove("voidedBy");
```
Comment (areklalo) [inline api/src/main/java/org/openmrs/module/sync2/api/utils/SyncHashcodeUtils.java]: Do we want to delete this information?

File: api/src/main/java/org/openmrs/module/sync2/api/utils/SyncHashcodeUtils.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+			result.remove("voidReason");
+		} else if (BaseOpenmrsMetadata.class.isAssignableFrom(clazz)) {
+			result.remove("dateRetired");
+			result.remove("retired");
```
Comment (areklalo) [inline api/src/main/java/org/openmrs/module/sync2/api/utils/SyncHashcodeUtils.java]: As above

File: api/src/main/java/org/openmrs/module/sync2/api/utils/SyncHashcodeUtils.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+	private static UUID getUuidFromString(String data) {
+		String md5Hex = DigestUtils.md5Hex(data);
+		md5Hex = md5Hex.replaceFirst(regex, replacement);
+		return UUID.fromString(md5Hex);
+	}
+
+	private static SimpleObject removeFields(SimpleObject simpleObject, Class<?> clazz) {
+		SimpleObject result = copySimpleObject(simpleObject);
+		result.remove("dateChanged");
```
Comment (areklalo) [inline api/src/main/java/org/openmrs/module/sync2/api/utils/SyncHashcodeUtils.java]: This field is wrapped by "auditInfo".

File: api/src/main/java/org/openmrs/module/sync2/api/utils/SyncHashcodeUtils.java (hunk 1 above)
Comment (wojtekbu) [reply api/src/main/java/org/openmrs/module/sync2/api/utils/SyncHashcodeUtils.java]: We have to detect if an object was voided on our instance, so we need to be able to tell if the received object was voided on the current instance. Adding voided to hashcode produce different hashcodes for objects with the same values.

**Pull request merged**